### PR TITLE
Prototype for PyTorch-specific flake8 plugin

### PR DESCRIPTION
**Pull request by kit1980**
The big vision for the plugin is to flag (and for some auto fix) all kinds of problematic usages of PyTorch.

What it is doing so far is printing error messages for the functions from the deprecated_symbols.yaml.
Eventually this list will be in the pytorch/pytorch and also will be used for generating documentation about deprecated and removed functions.

The plugin is in a working state, there is an output from checking torchaudio:

```
 flake8 --isolated --select=TOR
./examples/libtorchaudio/augmentation/create_jittable_pipeline.py:38:22: TOR101 Use of deprecated function torch.norm
./examples/tutorials/audio_data_augmentation_tutorial.py:193:14: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:475:16: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:477:30: TOR101 Use of deprecated function torch.norm
./test/torchaudio_unittest/transforms/transforms_test_impl.py:479:30: TOR101 Use of deprecated function torch.norm
./torchaudio/functional/functional.py:1039:27: TOR101 Use of deprecated function torch.norm
./torchaudio/functional/functional.py:1040:27: TOR101 Use of deprecated function torch.norm
```

### CODE REVIEW

File: tools/flake8-torch/flake8_torch/checker.py
Comment (ZainRizvi) [inline tools/flake8-torch/flake8_torch/checker.py:56]: this could be a good place to check for duplicate items with the same name

File: tools/flake8-torch/flake8_torch/checker.py
Comment (ZainRizvi) [inline tools/flake8-torch/flake8_torch/checker.py:51]: What is this a tree of, and where does it come from?  I bit of context on how linter plugins work would be helpful here

File: tools/flake8-torch/flake8_torch/checker.py
Comment (ZainRizvi) [inline tools/flake8-torch/flake8_torch/checker.py:90]: I'd love to hear your thoughts on what the rollout process would look like when someone wants to deprecate a new function once this feature is in.  I'm especially curious about when build breaks are expected to happen, when (potentially ignorable) warnings would be given, and when issues/tasks would be filed.

An idealized form might look something like:
1. Dev A marks function torch.Foo as deprecated. As part of his PR, he runs a script to grandfather in all existing usages of that deprecated function, blocking new invocations.
2. Once the PR is merged, issues/tasks get filed against the code owners to remove those now-deprecated usages. (Perhaps in the future we could even offer tooling that auto-replace the usages)
3. As the deprecated function's invocations are replaced, keep shrinking the list of places that are allowed to use that deprecated function (clean up the list)
4. Until the list of invocations has shrunk to zero, any PR that attempts to mark the function as removed will fail

**Review by ZainRizvi — COMMENTED** [review]
This is looking pretty promising!

I'd love to hear your thoughts on what the rollout process would look like when someone wants to deprecate a new function once this feature is in. 

I'm especially curious about when build breaks are expected to happen, when (potentially ignorable) warnings would be given, and when issues/tasks would be filed.

An idealized form might look something like:

1. Dev A marks function torch.Foo as deprecated. As part of his PR, he runs a script to grandfather in all existing usages of that deprecated function, blocking new invocations.
2. Once the PR is merged, issues/tasks get filed against the code owners to remove those now-deprecated usages. (Perhaps in the future we could even offer tooling that auto-replace the usages)
3. As the deprecated function's invocations are replaced, keep shrinking the list of places that are allowed to use that deprecated function (clean up the list)
4. Until the list of invocations has shrunk to zero, any PR that attempts to mark the function as removed will fail

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py:56]: Maybe later.

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py:51]: This is an AST tree, should be self-explanatory in this context.

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py:92]: I just came up with the codes, they are arbitrary.
The idea that the errors should be grouped logically in some way, so users can filter out classes of errors they are not interested in.

File: tools/flake8-torch/flake8_torch/deprecated_symbols.yaml
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/deprecated_symbols.yaml:7]: This file will eventually live in pytorch/pytorch, and there will be probably another checker on pytorch/pytorch looking for PRs with module:deprecation label or even search for strings "deprecated".
We'll need to discuss this.
This prototype is mostly to facilitate the discussion, but it's also already useful as is to find issues.

File: tools/flake8-torch/tests/fixtures/deprecated_from_nn.py
Comment (malfet) [inline tools/flake8-torch/tests/fixtures/deprecated_from_nn.py:8]: How about that?
```suggestion
import torch.nn as yy
yy.UpsamplingNearest2d()

import torch.nn.UpsamplingNearest2d as ZZZ
ZZZ()

FOO = ZZZ
FOO()
```

File: tools/flake8-torch/pyproject.toml
Comment (malfet) [inline tools/flake8-torch/pyproject.toml:1]: Please add dependency on `yaml`

File: tools/flake8-torch/flake8_torch/checker.py
Comment (malfet) [inline tools/flake8-torch/flake8_torch/checker.py:49]: Is it possible to keep versions in sync between pyproject.toml and this file?

File: tools/flake8-torch/flake8_torch/checker.py
Comment (malfet) [inline tools/flake8-torch/flake8_torch/checker.py:40]: Is `asname` guaranteed to be defined? If not, use `getattr`
```suggestion
            if getattr(alias, 'asname', None) is None:
```

File: tools/flake8-torch/flake8_torch/checker.py
Comment (malfet) [inline tools/flake8-torch/flake8_torch/checker.py:43]: One f-string to rull them all...
```suggestion
                self._import_info[alias.name] = f"{module}.{alias.name}"
            else:
                self._import_info[alias.asname] = f"{module}.{alias.name}"
```

**Review by malfet — APPROVED** [review]
LGTM, but how about adding a CI workflow that'll test those?
Also, please consider addressing nits

**Review by malfet — APPROVED** [review]
LGTM, but how about adding a CI workflow that'll test those?
Also, please consider addressing nits

File: tools/flake8-torch/tests/fixtures/deprecated_from_nn.py
Comment (kit1980) [reply tools/flake8-torch/tests/fixtures/deprecated_from_nn.py:8]: `FOO = ZZZ; FOO()` will not be caught currently.
I don't think it's a very common pattern.
We can add that later if needed, may be a bit tricky to implement, but should be doable.

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py:40]: `asname` is always defined, but can be None: https://docs.python.org/3/library/ast.html#ast.alias

File: tools/flake8-torch/tests/test_flake8_torch.py
Comment (kit1980) [reply tools/flake8-torch/tests/test_flake8_torch.py]: Removed that function, no longer needed.

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py]: Done

File: tools/flake8-torch/flake8_torch/checker.py
Comment (kit1980) [reply tools/flake8-torch/flake8_torch/checker.py]: done

**Comment by kit1980** [comment]
> LGTM, but how about adding a CI workflow that'll test those? Also, please consider addressing nits

I've updated the code to address the comments.
I'll add a workflow to run the tests in the next PR.

**Pull request merged**
